### PR TITLE
Add W&B experiment logger 

### PR DESCRIPTION
**Pull request by AyushExel**
## This PR adds:
* A unified base logger API 
* Use the new logger api to add support for existing mlflow logger and new W&B logger
* 
#### Describe the changes you've made
- [x] create a unified logging interface api (abstract class BaseLogger)

- [x] use that interface to refactor the existing mlflow integration

- [x] create wandb logger using the same interface..

- These external loggers will be called dashboard loggers and will be handled by a handler called DashboardLogger which can take a list of BaseLogger objects. Like `dash_logger = DashboardLogger([mlflowLogger(), wandbLogger(...)])`
- In order to make it backwards compatible  I'm allowing log_experiment to take 2 values..boolean or a string.. If true, use both loggers if available..use can also pass a string specifying wandb or mlflow to use a specific logger
- In order to improve the logging time Don't spin up a new run for each model being compared..Instead, initialize a run only at the beginning and keep logging to that run until user manually reinits..Unlike mlflow which treats each model separately

Some W&B logger features include:
* Logging interactive dataset/csv panels
* Data profile
* interactive model comparison tables
* dataset/model versioning
* experiment configuration management
* metric logging/tracking
* plot visualization

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

A clear and concise description of it.

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### Example classification dashboard
[classification dashboard](https://wandb.ai/cayush/classification_dashboard?workspace=user-cayush)

TODO:
* Customized tests
* Improve docs/comments
* clean up 

@moezali1 This is a major refactor that removes the existing mlflow logger in the new unified logger api. There's probably some more cleaning up to do but I want to start a review cycle now to make sure we're on the same page. A few things I'd like to confirm are:
* Are you okay with the proposed design of creating a unified logger for all 3rd party loggers?
* There are other things that we could automate like logging a complete EDA/ drift profile with  a single command based on this new logging api. Would love to know your opinion
*

### CODE REVIEW

**Comment by moezali1** [comment]
Hi, @AyushExel Thank you so much for working on this. I am tagging @Yard1 @tvdboom for technical review. Some of my initial comments are:

- Can we not add a new parameter in the `setup`, say `log_backend`, and use that to determine the external type. Whereas `log_experiment` remains a boolean parameter. Is that possible? 

- I am okay with the design as this leaves us with the ability to expand to more 3rd party loggers in the future. @Yard1 and @tvdboom to comment further on this.

- Can you please share an end-to-end Notebook that I can use to perform an experiment and use the W&B logger. I would love to try it myself.

As of now, the following parameters are available in the `setup` for logging perspective. Are you collecting input from any of this for the W&B logger?

image

Thanks

**Comment by moezali1** [comment]
@ngupta23 @Yard1 @tvdboom

**Comment by AyushExel** [comment]
@moezali1 hey thanks for the response. Regarding your point 1, I'm not using any additional parameters for wandb. I'm just using all the ones that you mentioned in the screenshot

**Comment by ngupta23** [comment]
> Can we not add a new parameter in the setup, say log_backend, and use that to determine the external type. Whereas log_experiment remains a boolean parameter. Is that possible?

I think that is a good idea. Users will use only 1 backend. By default it should be mlflow, but users should be able to swicth to w&b is needed.

- Also, need to add w&b to requirements-optional.
- Also, users should not have to install w&b by default to run pycaret, The code should hopefully be structured accordingly.

**Comment by AyushExel** [comment]
@ngupta23 cool. yep, it'll be optional. We can add a check that conditionally imports wandb if user sets that as logging backend.

**Comment by Yard1** [comment]
Hey, thanks a lot! I will take a look shortly

**Comment by AyushExel** [comment]
@Yard1 cool! thanks.. There are a few things needed for touch ups but I'd love to know if you agree with the proposed design. After this I'll add:
* Optional import of wandb.. Only try to import wandb if user set the logging backend to wandb
* Support nlp experiments. I've currently only implemented this for tabular but I guess it should be simple to expand that to nlp as well.

**Comment by morganmcg1** [comment]
hey all, here is the [colab I put together](https://wandb.me/pycaret) for our webinar yesterday with PyCaret. It installs pycaret from @AyushExel 's fork and shows what a suggested workflow might look like, just in case its helpful with any testing of the integration :)

File: pycaret/loggers/wandb_logger.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import os
```
Comment (Yard1) [inline pycaret/loggers/wandb_logger.py:1]: Can you format the file with black?

**Review by Yard1 — COMMENTED** [review]
This is looking good, I will update the branch to match the latest changes soon, then we can add some tests

File: pycaret/loggers/wandb_logger.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+
+    def log_sklearn_pipeline(self, prep_pipe, model):
+        pipeline = deepcopy(prep_pipe)
+        pipeline.steps.append(["trained_model", model])
+        pickle.dump(pipeline, open("pipeline.pkl","wb"))
+        art = wandb.Artifact("pipeline", type="model")
+        art.add_file("pipeline.pkl")
+        self.run.log_artifact(art)
+        os.remove("pipeline.pkl")
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply pycaret/loggers/wandb_logger.py:55]: yep will do

File: pycaret/loggers/wandb_logger.py (hunk 1 above)
Comment (AyushExel) [reply pycaret/loggers/wandb_logger.py:1]: sure thing

File: pycaret/loggers/wandb_logger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AyushExel) [reply pycaret/loggers/wandb_logger.py]: @Yard1 In all other places this same method is used for logging artifacts. Is tempfile to be used in all places?

File: pycaret/loggers/wandb_logger.py (hunk 1 above)
Comment (AyushExel) [reply pycaret/loggers/wandb_logger.py:1]: done

**Comment by AyushExel** [comment]
@Yard1 I've added support for the loggers in the nlp module and formatted the entire logger/ folder using black. The tests will also be automatically added if `wandb` is installed during the CI testing the way code is structured. Let me know if something else is needed

File: pycaret/loggers/wandb_logger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yard1) [reply pycaret/loggers/wandb_logger.py]: We can have it just here for now. We are in process of updating other parts of pycaret anyway.

File: pycaret/loggers/mlflow_logger.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            )
+        except:
+            logger.warning("Couldn't infer MLFlow signature.")
+            signature = None
+        if not _is_unsupervised(_ml_usecase):
+            input_example = (
+                data_before_preprocess.drop([target_param], axis=1).iloc[0].to_dict()
+            )
+        else:
+            input_example = data_before_preprocess.iloc[0].to_dict()
```
Comment (Yard1) [inline pycaret/loggers/mlflow_logger.py:120]: None of those are actually used, right? Should this code be commented out/removed?

**Comment by Yard1** [comment]
@AyushExel I have pushed a commit with some changes, PTAL and respond to comments - afterwards, it should be good to merge! Thanks for your work.

File: pycaret/loggers/wandb_logger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yard1) [reply pycaret/loggers/wandb_logger.py]: Also, I guess we can use this instead? https://docs.wandb.ai/ref/python/artifact#new_file

[the remark this answers is not in the document]
Comment (AyushExel) [reply pycaret/loggers/wandb_logger.py]: Cool. I changed it Thanks

**Comment by AyushExel** [comment]
@Yard1 Addressed your comments. Please verify

**Review by Yard1 — APPROVED** [review]
Looks good, thanks!

**Comment by charlotte0930** [comment]
Using pycaret.classfication.setup got keyerror dashboard.logger

**Comment by AyushExel** [comment]
@charlotte0930 can you provide a reproducible example or the complete traceback from your error?

**Comment by AyushExel** [comment]
@charlotte0930 nvm! I was able to repro

File: pycaret/internal/tabular.py [hunk 4]
```diff
@@ -794,7 +797,7 @@ def setup(
     # declaring global variables to be accessed by other functions
     logger.info("Declaring global variables")
     global _ml_usecase, USI, html_param, X, y, X_train, X_test, y_train, y_test, seed, prep_pipe, experiment__, fold_shuffle_param, n_jobs_param, _gpu_n_jobs_param, create_model_container, master_model_container, display_container, exp_name_log, logging_param, log_plots_param, fix_imbalance_param, fix_imbalance_method_param, transform_target_param, transform_target_method_param, data_before_preprocess, target_param, gpu_param, _all_models, _all_models_internal, _all_metrics, _internal_pipeline, stratify_param, fold_generator, fold_param, fold_groups_param, fold_groups_param_full, imputation_regressor, imputation_classifier, iterative_imputation_iters_param
-
+    global dashboard_logger
```
[the remark this answers is not in the document]
Comment (charlotte0930) [reply pycaret/internal/tabular.py:800]: > 

Thank you! The problem solved.

**Pull request merged**